### PR TITLE
[recipe, reward] feat: add Qwen-Image-Edit V1 NPU recipe with PickScore device placement

### PR DESCRIPTION
**Pull request by Sky-Trigger**
### What does this PR do?

> [!IMPORTANT]
> This PR depends on [PR #488](https://github.com/verl-project/verl-omni/pull/488)
> and should be merged only after it. The two changes were originally developed
> together, but were split into separate PRs to keep the TransferQueue metadata
> correctness fix and reward-worker accelerator placement clear and independently
> reviewable.

This PR adds an opt-in accelerator-aware local reward loop for the synchronous
V1 diffusion trainer and provides an Ascend NPU Qwen-Image-Edit-2511 FlowGRPO
recipe using PickScore.

When `reward.custom_reward_function.use_accelerator=True`, local reward workers
are placed into accelerator bundles from the existing actor/rollout resource
pool. For the checked-in single-node recipe, assignment starts at rank/bundle 0
and proceeds in order. For example, `reward.num_workers=4` places the workers on
ranks 0, 1, 2, and 3, with one local PickScore model instance per worker. The
manager rejects configurations requesting more reward workers than available
accelerator bundles.

The checked-in topology creates four TP=4 rollout replicas and 16 PickScore
reward workers. Each completed sample selects one worker at random from the
full set of 16; rollout tensor parallelism does not partition or bind the
reward worker set.

This is intentionally a temporary integration for Qwen-Image-Edit on the
synchronous V1 trainer. It only establishes explicit accelerator placement for
local custom reward workers. A broader reward-model lifecycle design, including
wake-up, sleep/offload, memory reclamation, and async coordination, is deferred
to follow-up work.

### Checklist Before Starting

- [x] Searched for similar PRs: [open PRs matching `accelerator reward worker`](https://github.com/verl-project/verl-omni/pulls?q=is%3Apr+is%3Aopen+accelerator+reward+worker).
- [x] PR title follows `[{modules}] {type}: {description}`:
  `[recipe, reward] feat: add Qwen-Image-Edit V1 NPU recipe with PickScore device placement`.

### Test

The checked-in NPU launcher was run successfully with its current training
hyperparameters in the author's environment. The model, reward-function, and
dataset paths are configurable; the validated training settings are preserved.

#### Training curve

`rewards/mean` over training steps:

2153915c5e2e1bce1e42e8b4fdcecc84

Focused CPU tests were run with
`/data/8T/wangmb/LLM-Infra/myenv/bin/python -m pytest` and aligned local source
checkouts:

```bash
PYTHONPATH=<this-verl-omni-checkout>:/data/8T/wangmb/LLM-Infra/verl:/data/8T/wangmb/LLM-Infra/vllm:<pinned-vllm-omni-checkout> \
    /data/8T/wangmb/LLM-Infra/myenv/bin/python -m pytest -q \
    tests/reward_loop/test_reward_loop_manager_on_cpu.py
```

```text
4 passed
```

The new PickScore platform-device test and its adjacent scoring test were also
run:

```bash
PYTHONPATH=<this-verl-omni-checkout>:/data/8T/wangmb/LLM-Infra/verl:/data/8T/wangmb/LLM-Infra/vllm:<pinned-vllm-omni-checkout> \
    /data/8T/wangmb/LLM-Infra/myenv/bin/python -m pytest -q \
    tests/utils/reward_score/test_pickscore_reward_on_cpu.py::test_inferencer_uses_platform_device_by_default \
    tests/utils/reward_score/test_pickscore_reward_on_cpu.py::test_score_encodes_duplicate_prompts_once_and_preserves_pairing
```

```text
2 passed
```

Additional checks:

```text
bash -n examples/flowgrpo_trainer/qwen_image_edit/run_qwen_image_edit_lora_v1_npu.sh  # passed
scripts/generate_trainer_config.sh                                                    # passed with pinned verl
python -m py_compile <changed Python files>                                         # passed
git diff --check                                                                    # passed
```

Full `pre-commit` and Ruff were not run because those tools are not installed
in the available environment.

### API and Usage Example

This PR adds one opt-in custom reward setting with a default value of `false`.
Existing reward-loop behavior is unchanged unless it is explicitly enabled.

```bash
reward.custom_reward_function.use_accelerator=True \
reward.num_workers=4 \
reward.reward_model.enable=False
```

The Qwen-Image-Edit NPU recipe can be launched with:

```bash
WORKSPACE=$PWD \
NUM_GPUS_ACTOR_ROLLOUT_REWARD=16 \
bash examples/flowgrpo_trainer/qwen_image_edit/run_qwen_image_edit_lora_v1_npu.sh
```

### Design & Code Changes

- Add `LocalAcceleratorRewardLoopManager` for opt-in local custom reward
  functions in the synchronous V1 diffusion trainer.
- Reuse the actor/rollout placement groups; in the checked-in single-node
  recipe, assign reward workers to accelerator bundles from rank 0 upward.
- Preserve the existing `OmniRewardLoopManager` path when accelerator placement
  is not requested.
- Reject incompatible standard reward-model usage and excessive local worker
  counts with explicit errors.
- Select PickScore's default device through the platform device helper instead
  of hard-coding CUDA.
- Add focused CPU coverage for bundle assignment, worker limits, default
  behavior, and PickScore device selection.
- Register `use_accelerator` in the reward schema with a default value of
  `false` and regenerate the reference configs.
- Add and document the validated Qwen-Image-Edit Ascend NPU recipe with the
  opt-in `reward.custom_reward_function.use_accelerator=True` override.

### Scope

This PR does not:

- include the Qwen-Image-Edit TransferQueue metadata fix from PR #488;
- change the separate async trainer;
- implement reward-model wake-up, sleep/offload, or distributed model
  lifecycle management;
- shard a worker-local PickScore model with tensor parallelism.

### CODE REVIEW

File: examples/flowgrpo_trainer/qwen_image_edit/README.md [hunk 1]
```diff
@@ -96,6 +96,30 @@ bash examples/flowgrpo_trainer/qwen_image_edit/run_qwen_image_edit_lora.sh \
...
+### Ascend NPU
+
+The NPU recipe uses the synchronous V1 diffusion trainer and assigns each
+local PickScore reward worker to an accelerator bundle from the actor/rollout
+resource pool:
+
+```bash
+WORKSPACE=$PWD \
+NUM_GPUS_ACTOR_ROLLOUT_REWARD=16 \
+bash examples/flowgrpo_trainer/qwen_image_edit/run_qwen_image_edit_lora_npu.sh
```
Comment (SamitHuang) [inline examples/flowgrpo_trainer/qwen_image_edit/README.md]: This command still points at `run_qwen_image_edit_lora_npu.sh`, but the second commit renamed the launcher to `run_qwen_image_edit_lora_v1_npu.sh`. The PR description and `bash -n` check use the old name as well, so the documented NPU entrypoint will fail as soon as someone copies it. Update the README, PR body, and any remaining references to the current filename.

**Review by SamitHuang — CHANGES_REQUESTED** [review]
Requesting changes. The CPU tests cover bundle indexing and the factory switch, but they do not validate remaining placement-group capacity, NPU device isolation, or colocated PickScore HBM. The NPU recipe path in the README is also broken after the launcher rename.

`verl_omni/reward_loop/local_accelerator_reward_loop.py:47` — `ray_resource_options(1 / max_colocate_count)` claims a slice of each actor/rollout bundle without checking how much that bundle already reserved. On the FSDP path `max_colocate_count` is often 1, so this can request a full accelerator on a PG that already hosts ActorRollout and hang in `_init_reward_loop_workers()`. The CPU test uses `max_colocate_count=3` only. Compare against remaining bundle capacity, or assign leftover ranks / a dedicated pool, and add a `max_colocate_count=1` test.

`verl_omni/utils/reward_score/pickscore_reward.py:54` — `get_device_name()` is the platform string (`npu`/`cuda`), not the bundle-local index. `model.to("npu")` lands on device 0 of whatever is visible. Fractional custom `NPU` resources do not necessarily set `ASCEND_RT_VISIBLE_DEVICES` the way `num_gpus=1` sets `CUDA_VISIBLE_DEVICES`. If isolation is incomplete, every worker loads PickScore onto npu:0. Bind to `get_device_id()` / `torch.npu.current_device()` after confirming Ray isolates the assigned NPU.

`examples/flowgrpo_trainer/qwen_image_edit/run_qwen_image_edit_lora_v1_npu.sh:13` — default `REWARD_WORKERS=16` places a resident PickScore CLIP on every actor/rollout NPU, including the TP=4 vLLM shards, with no sleep/offload. Ray fractions do not partition HBM. Either publish per-card HBM with actor+rollout+PickScore, or default to fewer workers off the vLLM TP group until lifecycle offload exists.

`run_qwen_image_edit_lora_v1_npu.sh:63` / `reward.yaml` — `use_accelerator` is not in the reward schema, so the recipe must use Hydra `+` and a normal override will fail struct-config validation. Add `use_accelerator: false` to `custom_reward_function` and drop the `+`.

Depends on #488 as stated; do not merge this first.

File: verl_omni/trainer/config/_generated_diffusion_trainer.yaml
Comment (zhtmike) [inline verl_omni/trainer/config/_generated_diffusion_trainer.yaml]: A little big strange you modified the four auto generated yaml.

**Comment by Sky-Trigger** [comment]
@SamitHuang  updated， PTAL. 

The `max_colocate_count=1` case, explicit device binding via `get_device_id()`, and the `use_accelerator` schema issue have all been fixed.

Regarding the HBM concern, this change does not make the current memory situation worse. In fact, it is already an improvement over the previous behavior, where all PickScore models could end up being loaded onto device 0. For now, the PickScore models are at least distributed across their assigned devices instead of being concentrated on a single device.

I will follow up shortly with a dedicated lifecycle design for the reward models, including explicit offloading.

File: verl_omni/reward_loop/local_accelerator_reward_loop.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+# limitations under the License.
+
+from ray.util.scheduling_strategies import PlacementGroupSchedulingStrategy
+from verl.plugin.platform import get_platform
+from verl.utils.device import get_device_name
+
+from .reward_loop import OmniRewardLoopManager
+
+
+class LocalAcceleratorRewardLoopManager(OmniRewardLoopManager):
```
Comment (zhtmike) [inline verl_omni/reward_loop/local_accelerator_reward_loop.py:24]: Add a TODO here, mentioning it is `a temporary integration for Qwen-Image-Edit on the
synchronous V1 trainer`. SO other pipeline will not rely on it

**Review by zhtmike — APPROVED** [review]

File: verl_omni/reward_loop/local_accelerator_reward_loop.py (hunk 2 above)
Comment (Sky-Trigger) [reply verl_omni/reward_loop/local_accelerator_reward_loop.py:24]: updated

**Comment by zhtmike** [comment]
remember to revert these code changes once a formal solution is provided.

track in #445

**Comment by Sky-Trigger** [comment]
@SamitHuang  PTAL

**Comment by Sky-Trigger** [comment]
@zhtmike Could you please dismiss the requested changes, or should I wait a bit longer?

**Comment by zhtmike** [comment]
> @zhtmike Could you please dismiss the requested changes, or should I wait a bit longer?
> 
> 

Need @SamitHuang 's feedback.

**Review by SamitHuang — COMMENTED** [review]
LGTM

**Review by SamitHuang — APPROVED** [review]
